### ISSUE TITLE
Support clarifying SWRResponse type to do type narrowing SWRResponse["data"] by flags(isSuccess) or union type or any other ways please

### ISSUE THREAD

**Issue by manudeli** [vercel/swr#2482]
How can we use useSWR like below to do type narrowing for data?

```tsx
const SWRComp = () => {
  const swrQuery = useSWR(...)

  if(swrQuery.data !== undefined){
    return <>{swrQuery.data}</>
  }
}
```

but in this case, if fetcher to put in useSWR return undefined, how can we do guarantee fetcher's returning promise is success? you can check it this edge case. data is never.

![image](https://user-images.githubusercontent.com/61593290/223355180-8f9958d0-4bb2-4713-9f38-3e6bbcf03cc3.png)

I'm sad that there is no isSuccess flag in useSWR because of this problem.

Could you make it TypeScript example to guide us avoiding this problem?

In this example(https://swr.vercel.app/examples/basic) in official guide in JavaScript file. I can understand it because of JavaScript will accept it. but in TypeScript is not!🥲

### In JavaScript, no type error

```jsx
import React from "react";
import useSWR from "swr";

const fetcher = (url) => fetch(url).then((res) => res.json());

export default function App() {
  const { data, error, isLoading } = useSWR(
    "https://api.github.com/repos/vercel/swr",
    fetcher
  );

  if (error) return "An error has occurred.";
  if (isLoading) return "Loading...";
  return (
    <div>
      <h1>{data.name}</h1>
      <p>{data.description}</p>
      <strong>👁 {data.subscribers_count}</strong>{" "}
      <strong>✨ {data.stargazers_count}</strong>{" "}
      <strong>🍴 {data.forks_count}</strong>
    </div>
  );
}

```

### In TypeScript, Type error will be occured
```tsx
import React from 'react'
import useSWR from 'swr'

const fetcher = (
  url: string
) => fetch(url).then((res) => res.json()) as Promise<{
  name: string
  description: string
  subscribers_count: number
  stargazers_count: number
  forks_count: number
}>

export default function App() {
  const { data, error, isLoading } = useSWR(
    'https://api.github.com/repos/vercel/swr',
    fetcher
  )

  if (error) return 'An error has occurred.'
  if (isLoading) return 'Loading...'
  return (
    <div>
      <h1>{data.name}</h1>
      <p>{data.description}</p>
      <strong>👁 {data.subscribers_count}</strong>{' '}
      <strong>✨ {data.stargazers_count}</strong>{' '}
      <strong>🍴 {data.forks_count}</strong>
    </div>
  )
}

```

<img width="585" alt="image" src="https://user-images.githubusercontent.com/61593290/223337008-6fb0c6a5-4a72-447f-abfe-19d49c75ebb6.png">

How do we have to avoid this type error situation?

_Originally posted by @manudeli in https://github.com/vercel/swr/issues/2175#issuecomment-1457608985_
            

### What I expected: union return type of useSWR containing flag expressing Promise's status
in @tanstack/react-query, they support union return type of useQuery, so we can do type narrowing by isSuccess, isLoading, isError.
but in swr, it's not.

#### type narrowing by isSuccess in @tanstack/react-query
```tsx
const ReactQueryComp = () => {
  const reactQuery = useQuery(...)

  if (reactQuery.isSuccess){ // even if data is undefined, when Promise fulfilled, we can use it! not like swr
    return <>{reactQuery.data}</>
  }
}
```

#### type narrowing by isLoading, isError in @tanstack/react-query
```tsx
const ReactQueryComp = () => {
  const reactQuery = useQuery(...)

  if (reactQuery.isLoading) return <>laoding...</>
  if (reactQuery.isError) return <>{reactQuery.error}</> 
  return <>{reactQuery.data}</> // data type narrowed by isLoading, isError
}
```

**Comment by tobbbe** [comment]
I dont think we need isSuccess? Because if x.error has value or x.isValidating/x.isLoading is true. Then we can assume isSuccess is false right?

**Comment by manudeli** [comment]
> I dont think we need `isSuccess`? Because if `x.error` has value or `x.isValidating`/`x.isLoading` is `true`. Then we can assume `isSuccess` is false right?

Then don't we need to provide `data` type narrowing in this below picture? but why data can be undefined typescriptly?
Then you mean you want to make SWRResponse do type narrowing `data` by only error, isValidating, isLoading without isSuccess flag? like below picture?

### type narrowing by isLoading, error
<img width="826" alt="image" src="https://user-images.githubusercontent.com/61593290/223372570-0926c5e3-6305-40c6-9381-2b0fa82ee190.png">

### type narrowing by isLoading, error, isValidating
<img width="812" alt="image" src="https://user-images.githubusercontent.com/61593290/223371740-e5a964a6-c40c-47a8-9dbb-d3d8aaaa2473.png">

I expected their `data` type  need to be just `"success"` without undefined. anyway in this case too without isSuccess flag.

**Comment by tobbbe** [comment]
Yea true! `x.data` can be both `undefined` and `some value` while isValidating.

Im not sure why data can be undefined in that case 🤔 useSWR hook must do something. Ill keep looking!

**Comment by mnik01** [comment]
facing a bug
<img width="760" alt="image" src="https://user-images.githubusercontent.com/39025497/229359661-d2a08daa-8037-4f14-882b-c75cb7ae444c.png">
<img width="484" alt="image" src="https://user-images.githubusercontent.com/39025497/229359670-bdad2cc8-71ab-47c8-9bfb-b3e878732acd.png">


In data is possibly undefined. But fetcher returns a Promise<User> which is object not undefined

**Comment by marcelocarmona** [comment]
@tobbbe @manudeli I'm interested in this feature I like what I see in https://github.com/vercel/swr/pull/2175 Do you know if there is something missing here or edge cases that are not covered?

Also for more context about this feature, this is a good comparison with react-query https://news.ycombinator.com/item?id=33929553

**Comment by lcswillems** [comment]
TanStack Query implements this out of the box: https://tanstack.com/query/latest/docs/framework/react/typescript#type-narrowing

swr is a no-go for me just because of this.
